File: ProblemSolve/16.cpp
/****************
 * 3Sum Closest *
 ****************/

// 참고 : https://leetcode.com/problems/3sum-closest/discuss/7883/C%2B%2B-solution-O(n2)-using-sort

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int len = nums.size() - 1 ;
        sort(nums.begin(), nums.end());        
        int closest = nums[0] + nums[1] + nums[2];
        int second, third;
                
        for(int first = 0; first < len-1; first++){            
            second = first + 1;
            third = len;
            
            while(second < third){
                int currSum = nums[first] + nums[second] + nums[third];
                if(currSum == target){
                    return currSum;
                }
                
                if(abs(target - currSum) < abs(target - closest)){
                    closest = currSum;
                }
                
                if(currSum > target){
                    third--;
                }
                else {
                    second++;
                }
            }            
        }        
        
        return closest;        
    }
};

File: ProblemSolve/16.cs
/****************
 * 3Sum Closest *
 ****************/

public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int closest = nums[0] + nums[1] + nums[nums.Length-1];
        
        for(int i=0; i<nums.Length-2; ++i){
            int left = i + 1;
            int right = nums.Length - 1;
            
            while(left < right){
                int currSum = nums[i] + nums[left] + nums[right];
                
                if(currSum == target){
                    return currSum;
                }                                
                
                if(Math.Abs(target - closest) > Math.Abs(target - currSum)){
                    closest = currSum;
                }
                
                if(currSum > target){
                    right--;
                }
                else{
                    left++;
                }
            }
        }
        
        return closest;
    }
}